### PR TITLE
Add distillation script for faster inference

### PR DESCRIPTION
**Pull request by Flova**
## Proposed changes

This adds an experimental script that is able to distill a diffusion model into one that only does a single step.

## Checklist

- [ ] Write documentation
- [x] Create issues for future work
- [x] This PR is on our `DDLitLab` project board

### CODE REVIEW

File: ddlitlab2024/dataset/pytorch.py [hunk 1]
```diff
@@ -40,15 +39,15 @@ class DDLITLab2024Dataset(Dataset):
     @dataclass
     class Result:
         joint_command: torch.Tensor
-        joint_command_history: torch.Tensor
-        joint_state: torch.Tensor
-        rotation: torch.Tensor
-        game_state: torch.Tensor
-        image_data: torch.Tensor
-        image_stamps: torch.Tensor
+        joint_command_history: Optional[torch.Tensor]
```
Comment (texhnolyze) [inline ddlitlab2024/dataset/pytorch.py:42]: While [https://peps.python.org/pep-0604/](equivalent) `type | None` is recommended for python 3.10+ https://peps.python.org/pep-0604/

Comment (texhnolyze) [reply ddlitlab2024/dataset/pytorch.py:42]: I'm a little confused why `joint_command` is not optional, but everything else is.
Is this `Result` used in two different ways, where maybe two diffferent data objects would make more sense?

File: ddlitlab2024/dataset/pytorch.py [hunk 2]
```diff
@@ -339,12 +366,14 @@ def __getitem__(self, idx: int) -> Result:
     def collate_fn(batch: Iterable[Result]) -> Result:
         return DDLITLab2024Dataset.Result(
             joint_command=torch.stack([x.joint_command for x in batch]),
-            joint_command_history=torch.stack([x.joint_command_history for x in batch]),
-            joint_state=torch.stack([x.joint_state for x in batch]),
-            image_data=torch.stack([x.image_data for x in batch]),
-            image_stamps=torch.stack([x.image_stamps for x in batch]),
-            rotation=torch.stack([x.rotation for x in batch]),
-            game_state=torch.tensor([x.game_state for x in batch]),
+            joint_command_history=torch.stack([x.joint_command_history for x in batch])
```
Comment (texhnolyze) [inline ddlitlab2024/dataset/pytorch.py:369]: I would always prefer the usage of empty collections instead of `null|None` whenever possible to prevent issues with NullPointers and be able to use consistent collection APIs without needing to rely on null checks.
I'm unsure of the context here though. Would an empty `torch.stack([])` break training?

File: ddlitlab2024/ml/model/encoder/base.py [hunk 3]
```diff
@@ -9,18 +9,22 @@ class BaseEncoder(nn.Module):
...
+        :param patch_size: The size of the patches for the convolutional embedding.
         :param hidden_dim: The number of hidden dimensions.
         :param num_layers: The number of transformer layers.
         :param num_heads: The number of attention heads.
         :param max_seq_len: The maximum length of the input sequences (used for positional encoding
         """
         super().__init__()
-        self.embedding = nn.Linear(input_dim, hidden_dim)
+        # Embed into non-overlapping patches
+        self.embedding = nn.Conv1d(input_dim, hidden_dim, kernel_size=patch_size, stride=patch_size)
```
Comment (texhnolyze) [inline ddlitlab2024/ml/model/encoder/base.py:27]: Can you explain how this embedding is better than the linear one?

File: ddlitlab2024/dataset/pytorch.py [hunk 4]
```diff
@@ -178,25 +194,36 @@ def query_image_data(
...
+        )
+
         # Get the raw image data
         for stamp, data in response:
             # Deserialize the image data
             image = np.frombuffer(data, dtype=np.uint8).reshape(480, 480, 3)
-            # Make chw from hwc
-            image = np.moveaxis(image, -1, 0)
+            # Resize the image
+            image = cv2.resize(image, (resolution, resolution), interpolation=cv2.INTER_AREA)
```
Comment (texhnolyze) [inline ddlitlab2024/dataset/pytorch.py:211]: Do we do this resize for faster training / less VRAM usage? I assume in our DB all images are still 480x480 or did that change for the NAO images?

File: ddlitlab2024/ml/training/train.py [hunk 5]
```diff
@@ -147,7 +177,7 @@
     )
 
     # Load the learning rate scheduler state if a checkpoint is provided
-    if args.checkpoint is not None:
+    if args.checkpoint is not None and False:
```
Comment (texhnolyze) [inline ddlitlab2024/ml/training/train.py:180]: This will always result in `False`.
I assume you meant `if args.checkpoint is not None and args.checkpoint is not False` in which case I would just do `if args.checkpoint:`

But I don't really see when it could be False anyway, since it is defined as a `str` argument.

**Review by texhnolyze — CHANGES_REQUESTED** [review]

File: ddlitlab2024/dataset/pytorch.py (hunk 4 above)
Comment (Flova) [reply ddlitlab2024/dataset/pytorch.py:211]: I added this so we can train using a lower resolution. Some models are e.g. pretrained with 224*224 so it is best that we use this resolution. It also helps with vram. The resolution in the DB would serve as our maximum resolution.

File: ddlitlab2024/ml/training/train.py (hunk 5 above)
Comment (Flova) [reply ddlitlab2024/ml/training/train.py:180]: It was just a hack to deactivate this code path. I will clean this up with another flag. Because loading the learning rate scheduler state when starting from a pretrained model is not desirable (as the end of the schedule was reached during pretrained). In contrast to that you want to resume the schedule if e.g. the training was interrupted.

**Review by texhnolyze — APPROVED** [review]
